### ISSUE TITLE
style and className props aren't being correctly reconciled with the DOM during hydration when the client generates different values than the server

### ISSUE THREAD

**Issue by justincy** [vercel/next.js#17463]
# Bug report

## Describe the bug

During hydration, when the client generates code which doesn't match what the server generated, the `className` and `style` props from the server are persisting instead of being updated to reflect what the client generated.

## To Reproduce

Here's a sample app the reproduces the behavior: https://github.com/justincy/debug-nextjs-ssr-styles

## Expected behavior

During hydration, when the client generates code which doesn't match what the server generated, I expect the DOM to update to reflect what the client generated. This correctly happens with text but isn't happening with the `className` and `style` props.

## System information

- OS: macOS
- Browser: Chrome
- Version of Next.js: 9.5.3
- Version of Node.js: 12.18.2

## Additional context

If I put a `debugger` statement right before the `return` from the `Home` page and follow the rendering path, it appears that the VDOM is out of sync with the DOM. The VDOM reflects what the client is generating so it doesn't commit any changes to the DOM.

I discovered this behavior while working with media queries. They can't be correctly processed on the server so during hydration I expect them to be updated and potentially modify the layout. I'd love to know whether I'm doing some wrong and can avoid this bug.

**Comment by timneutkens** [comment]
This is related to the way React hydration works. It's expected that you get unexpected results when hydration mismatches. This generally means that you're doing calculations during rendering that should happen in a `useEffect` instead. In case of the example repository that's definitely the case. I believe there's a few long threads on the React repository that you could look at with regards to hydration mismatches.

**Comment by justincy** [comment]
Thank you @timneutkens. I googled my heart out yesterday for solutions to my problem but I think I was being too specific in my search terms. `hydration mismatches` turned up useful results. I found a solution to my problem. I'll document some of my findings here in case it's useful for others in the future. The behavior is very unexpected, but makes sense when you dive into it and see that it's intentional for performance reasons.

* This is expected behavior according to the docs about [React.hydrate()](https://reactjs.org/docs/react-dom.html#hydrate).
* Here's a good [GitHub issues thread from react about hydration mismatches](https://github.com/facebook/react/issues/10591)
* Josh W Comeau wrote a thorough article about this issue titled [The perils of rehydration](https://joshwcomeau.com/react/the-perils-of-rehydration/). It's from the perspective of Gatsby but applies to any SSR of React, including Next.js.

I ended up solving my problem with the `useHasMounted()` hook from Josh Comeau's article. This enables two-pass rendering: first pass on the server, I don't apply the inline styles; for the second pass, a rerender is triggered on the client after hydration, at which point I apply the styles. For me, it looks like this:

```jsx
function Component({ children }) {
    const hasMounted = useHasMounted();
   // Generate some inline styles here...
   const styles = {...};
  return <div style={hasMounted ? styles : {}}>{children}</div>
}
```
